### PR TITLE
HW4 - Add new functionality, interfaces, abstract classes

### PR DESCRIPTION
**Pull request by artyomshpl**
HW 4. I'm not sure if I can rename the classes Ticket and others by adding the suffix Impl, because in that case, the code will not fully comply with the technical assignment requirements.

### CODE REVIEW

File: src/main/java/com/shep/fourthlecture/models/Ticket.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import lombok.Getter;
+import lombok.Setter;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+
+@Getter
+@Setter
```
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/models/Ticket.java:17]: cannot be for all fields

File: src/main/java/com/shep/fourthlecture/models/Ticket.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+public class Ticket implements Identifiable, Printable {
+    @NotNull
+    private int id;
+    private String concertHall;
+    private String eventCode;
+    private Long time;
+    private boolean isPromo;
+    private char stadiumSector;
+    private double maxBackpackWeight;
+    private double price;
```
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/models/Ticket.java:34]: type is not correct

File: src/main/java/com/shep/fourthlecture/models/Ticket.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                            + " Your value: " + id + " is " + idForCheck.length());
+                }
+            }
+        } catch (IllegalArgumentException e) {
+            System.out.println(e.getMessage() + " in Ticket: " + this);
+        }
+    }
+
+
+    public static LocalDateTime convertUnixTimestampToLocalDateTime(long unixTimestamp) {
```
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/models/Ticket.java]: better to do it in the specific, util class, not in the model

File: src/main/java/com/shep/fourthlecture/services/TicketService.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import com.shep.fourthlecture.sharing.PhoneShare;
+import com.shep.fourthlecture.sharing.impl.Shareable;
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
```
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/services/TicketService.java]: why do you use it?

File: src/main/java/com/shep/fourthlecture/models/base/User.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    @Override
+    public void setId(int id) {
+        this.id = id;
+    }
+
+    public void printRole() {
+        System.out.println("Role: " + role);
+    }
+
+    public abstract void uniqueFunction(Ticket ticket);
```
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/models/base/User.java]: incorrect name

**Review by VolgaDaniluk — COMMENTED** [review]
task with sharing doesn't complete

File: src/main/java/com/shep/fourthlecture/models/Ticket.java (hunk 2 above)
Comment (artyomshpl) [reply src/main/java/com/shep/fourthlecture/models/Ticket.java:27]: I'd like to clarify. The type is not correct in price field, right?

File: src/main/java/com/shep/fourthlecture/models/base/User.java
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/models/base/User.java:1]: you have 2 getter and setter for id :)

File: src/main/java/com/shep/fourthlecture/models/base/User.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    @Override
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public void printRole() {
+        System.out.println("Role: " + role);
+    }
+
+    public abstract void handleTicket(Ticket ticket);
```
Comment (VolgaDaniluk) [inline src/main/java/com/shep/fourthlecture/models/base/User.java]: "In your application, there should be 2 Users - Client and Admin, both should be able to .printRole() and also have unique functions - getTicket() for User and
checkTicket() - for Admin."

nothing about handleTicket

**Review by VolgaDaniluk — COMMENTED** [review]
formatting

**Pull request merged**